Lib9c/Policy/BlockPolicySource.Utils.cs

```diff
@@ -0,0 +1,380 @@
+            if (index == 0)
+            {
+                return true;
+            }
+            else if (IsAuthorizedMiningBlockIndex(blockChain, index))
+            {
+                return IsAuthorizedToMine(blockChain, miner, index);
+            }
+            else if (isPermissionedMiningBlockIndex(blockChain, index))
+            {
+                return isPermissionedToMine(blockChain, miner, index);
+            }
+
+            // If none of the conditions apply, any miner is allowed to mine.
+            return true;
+        }
+
+        internal static Func<BlockChain<NCAction>, Address, long, bool> IsAllowedToMineFactory(
+            Func<BlockChain<NCAction>, long, bool> isPermissionedMiningBlockIndex,
+            Func<BlockChain<NCAction>, Address, long, bool> isPermissionedToMine)
+        {
+            return (blockChain, miner, index) =>
+                IsAllowedToMineRaw(
+                    blockChain, miner, index, isPermissionedMiningBlockIndex, isPermissionedToMine);
+        }
+
+        /// <summary>
+        /// Checks if authorized mining policy applies to given block index.
+        /// </summary>
+        /// <remarks>
+        /// An implementation should be agnostic about other policies affecting the same index.
+        /// Policy overruling between different policies should be handled elsewhere.
+        /// </remarks>
+        internal static bool IsAuthorizedMiningBlockIndex(
+            BlockChain<NCAction> blockChain, long index)
+        {
+            if (GetAuthorizedMinersState(blockChain) is AuthorizedMinersState ams)
+            {
+                return index % ams.Interval == 0 && index <= ams.ValidUntil;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Checks if given miner is allowed to mine a block with given index according to
+        /// authorized mining policy.
+        /// </summary>
+        /// <remarks>
+        /// An implementation should be agnostic about other policies affecting the same index.
+        /// Policy overruling between different policies should be handled elsewhere.
+        /// </remarks>
+        internal static bool IsAuthorizedToMine(
+            BlockChain<NCAction> blockChain, Address miner, long index)
+        {
+            if (IsAuthorizedMiningBlockIndex(blockChain, index))
+            {
+                if (GetAuthorizedMinersState(blockChain) is AuthorizedMinersState ams)
+                {
+                    return ams.Miners.Contains(miner);
+                }
+                else
+                {
+                    throw new ArgumentException(
+                        $"Result of {nameof(GetAuthorizedMinersState)} cannot be null.");
+                }
+            }
+            else
+            {
+                throw new ArgumentException(
+                    $"Result of {nameof(IsAuthorizedMiningBlockIndex)} must be true.");
```

Comment:
It seems that checking the block index in this function is too much work.

Or, I think it's okay to return `false` here instead of raising `ArgumentException`. (then, we can remove branch [here](https://github.com/planetarium/lib9c/pull/640/files#diff-0e19782ab5996884e5ecd2b582cce259a2691938e566b1e7f09557abc8bea8a0R238-R241))

Reply:
The method is under the assumption that `IsAuthorizedMiningBlockIndex()` is `true` in the first place, so index check is not strictly necessary. This is purely for clarity and/or defensive reasons. I agree the code is pretty icky at the moment, but I failed to come up with a better solution.

- Asking whether a miner is authorized to mine on a non-target block index is non-sensical.
  - Can be answered either way as vacuously `true` or vacuously `false`, but this seemed dangerous.
- I wanted to have `IsAuthorizedToMine()` as a separate method, but perhaps not having it as a separate method may be better.
- Using ternary logic may be most natural with `bool?` as its return type. Wasn't sure it's a good practice to do so.
- This is most likely to get overhauled in a future PR when variable sub-policies feature finally gets implemented, so I didn't want to spend *too* much time on this issue. 🙃

Reply:
> Can be answered either way as vacuously true or vacuously false, but this seemed dangerous.

- Basically, I think it's good to ask for a simple case if possible and to use a combination of these when necessary. 
- This predicate appears to be a subfunction that is only called from `IsAllowedToMineRaw()`. If so, it just looks up if there is a miner (miner) corresponding to the state obtained with `GetAuthorizedMinersState` and returns the result. (If the name of this predicate was `BelongsAuthorizedMinerState()`, it wouldn't look too dangerous without index checking.)
  - Of course, we may have questions like "is such a predicate really necessary?", and I agree with that. If so, we can define its return as just `true` or `false` (no exception). and abstract `IsAllowedToMineRaw()` as like...
  ```csharp
          internal static bool IsAllowedToMineRaw(
            BlockChain<NCAction> blockChain,
            Address miner,
            long index,
            Func<BlockChain<NCAction>, long, bool> isPermissionedMiningBlockIndex,
            Func<BlockChain<NCAction>, Address, long, bool> isPermissionedToMine)
        {
            // For genesis blocks, any miner is allowed to mine.
            if (index == 0)
            {
                return true;
            }

            return IsAuthorizedToMine(blockChain, miner, index) || isPermissionedToMine(blockChain, miner, index); // ???
        }
  ```
  - Oh, it's a matter. we can't express vanilla state (no auth target and/or no permissioned target block) well. 
  - Maybe instead of a bool-based predicate, it makes more sense to use a checker-based approach that raises an exception... but it would be a lot of work to fix it all. understand.